### PR TITLE
add TryLSDGateway contract + test + deploy script

### PR DESCRIPTION
**Pull request by 0xEviso**

### CODE REVIEW

File: src/contracts/TryLSDGateway.sol [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    function swapAndDeposit(
+        address owner,
+        uint256 minShares
+    ) public payable returns (uint256 shares) {
+        // should send more than 0 eth
+        if (msg.value == 0) revert TooLittleEthError();
+
+        uint256 singleSwapAmount = msg.value / 3;
```
Comment (danoctavian) [inline src/contracts/TryLSDGateway.sol:173]: We'll get a dust amount left in the contract of 1 or 2 wei (msg.value % 2) depending on msg.value, we can attach that to the last transaction to keep it clean.

File: src/contracts/TryLSDGateway.sol [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+    // failed to transfer eth back to user after withdraw and swap
+    error FailedToSendEth();
+
+    /*//////////////////////////////////////////////////////////////
+                            EXTERNAL CONTRACTS
+    //////////////////////////////////////////////////////////////*/
+
+    // eth mainnet wsteth
+    IWstETH internal _wsteth =
```
Comment (danoctavian) [inline src/contracts/TryLSDGateway.sol:69]: All hardcoded things and things that are set in the constructor should be immutable to save gas 

```    IWstETH internal immutable _wsteth =
        IWstETH(0x7f39C581F595B53c5cb19bD0b3f8dA6c935E2Ca0);```

File: src/contracts/TryLSDGateway.sol [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        // calculate the amount of eth we get for singleSwapAmount of sfrxeth
+        // for get_dy asset 0 is eth, asset 1 is frxeth
+        ethAmount += _ethToFrxeth.get_dy(
+            1,
+            0,
+            _sfrxeth.convertToAssets(sfrxethAmount)
+        );
+    }
+
+    function withdrawAndSwap(
```
Comment (danoctavian) [inline src/contracts/TryLSDGateway.sol:250]: From a product perspective it's unclear to me this is something the users wants (unroll the position back into ETH).

Would consider not having this in this first release

File: src/test/unit/TryLSDGateway.t.sol [hunk 4]
```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: MIT
```
Comment (danoctavian) [inline src/test/unit/TryLSDGateway.t.sol:1]: Since this is a fork test we should rename directory to fork.

Generally, making /fork /unit and /integration directories for each category works well

File: src/contracts/TryLSDGateway.sol [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        uint256 sfrxethAmount = _sfrxeth.convertToShares(frxethAmount);
+
+        // finally calculate the amount of pool shares we get for the 3 tokens
+        shares = _tryLSD.calc_token_amount(
+            [wstethAmount, rethAmount, sfrxethAmount],
+            true
+        );
+    }
+
+    function swapAndDeposit(
```
Comment (danoctavian) [inline src/contracts/TryLSDGateway.sol:166]: As discussed off-band we can consider adding at some point a function that swaps through rETH when amounts are small and gas cost relative to size of transaction is significant and reduce the gas cost significantly by eliminating 2 of the swaps

File: src/contracts/TryLSDGateway.sol [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        _reth.approve(address(_ethToReth), type(uint256).max);
+        // unlimited approve will be used to swap frxeth to eth
+        _frxeth.approve(address(_ethToFrxeth), type(uint256).max);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            PAYABLE LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    bool _startedWithdraw;
```
Comment (danoctavian) [inline src/contracts/TryLSDGateway.sol:123]: I'd keep all variables in sections up top

* constants
* vars

While here it's less of an issue and it would just follow a standard, when dealing with contracts that are upgradeable proxies, memory layout needs to be preserved so having it all in 1 place is necessary.

**Comment by danoctavian** [comment]
We should fix the failing lint to have a perfectly green build

**Comment by danoctavian** [comment]
Screenshot 2023-12-03 at 4 13 45 AM


Gas reports generated for functions.

Implemented the REth only function  here for comparison:

https://github.com/yieldnest/trylsd-contracts/blob/867c32d341866d999d0de737bacb1fa8038773bc/src/contracts/TryLSDGateway.sol#L218

**Comment by danoctavian** [comment]
dust fix looks good few cleanups left that need to be done
 
 * immutable vars
 * grouping vars in 1 place
 * caliing it a fork test directory instead of unit
 * adding that extra test case
 * remove console log
 * fix lint

**Comment by 0xEviso** [comment]
@danoctavian  Can you expand a little more on :
- grouping vars in 1 place
- adding that extra test case
Thank you in advance

**Comment by danoctavian** [comment]
@0xEviso 

```bool internal _startedWithdraw;``` if we can put this at the top section after the immutables  to have all variables in 1 place


For the test case - testing the extreme case with very large amounts relative to pool size (deposit and withdraw) like this test:

https://github.com/yieldnest/trylsd-contracts/blob/temp/trylsdgateway-experiment/src/test/unit/TryLSDGateway.t.sol#L102

or if you can think of other edge case scenarios
